### ISSUE TITLE
Improve the listing of questions in a tag

### ISSUE THREAD

**Issue by skiadas** [skiadas/HanoverEvals#128]
They should look less spread out, and one should be able to reorder them. They should also probably be using our new typeahead system.

**Comment by altermattw** [comment]
I agree.  I think a tabular view, like we earlier considered for questions (Ref #136), would work well here.

**Comment by altermattw** [comment]
What do you think of something like [this](http://designpoodles.com/avtex/sortable-table/)? It's drag-and-droppable.  Uses jQueryUI.

**Comment by skiadas** [comment]
Looks good.

**Comment by altermattw** [comment]
The form currently works great except for one problem:  when you Add more questions than you started with (e.g., if you start with 6 and add 2 more), the excess questions in the collection are given the label "__new__", which is coming from the prototype.  This occurs not necessarily with newly added items.  For example, you can add 2 new items, then drag them to the top, and they will save perfectly and their new labels will print fine when the screen refreshes after the save.  The problem is with the questions that go beyond the starting number.  The last 2 questions have labels of "__new__".  The labels in the form itself are fine, so I'm thinking the problem may be in preparing Doctrine to receive new  questions.  Maybe there is some clue [here](http://symfony.com/doc/current/form/form_collections.html#allowing-new-tags-with-the-prototype)?

**Comment by altermattw** [comment]
Okay, the form is actually not persisting.  When you click "Save", you get the changes you expect (except the `__new__` issue, explained above), but when you back out to the Tags menu and then go back to edit, the changes disappear.

**Comment by skiadas** [comment]
Hm yeah something feels a bit weird there. So first of all, the whole thing doesn't seem to work if there are no questions in the tag to begin with; the add button does nothing.

The other thing that strikes me is that when I add a question, it looks very different because of the "select" menu that seems to stay with it.

I think the select should not be part of the prototype. When someone clicks to add a question, then they should somehow be presented with the list of questions, and choose one to add. And *then* that one question is used in conjunction with the prototype to create the entry to the list.  But the select of the questions is a separate entity, that just lives on the page. Does that make sense?

For persisting the forms, we'll almost certainly need to change the code in the controller somewhat, though I'm not quite sure what the right thing would be right now.

**Comment by skiadas** [comment]
@altermattw We should probably be using a "typeahead" for the "add question" functionality.

**Comment by altermattw** [comment]
I was thinking the same thing about adding questions:  If we have more than about 10 questions, the `<select>` strategy isn't going to look very good.  I was thinking that, rather than taking the user to a new page to pick out a question, we could insert a large `<div>` with some search functionality in it, which then shrinks after they make their selection.  Typeahead is a good idea.  Maybe the possible questions could also be presented in a table?

**Comment by skiadas** [comment]
It would work pretty much the same way we have typeahead working for adding professors to section. The key infrastructure is the same. There is a typeahead element, and a collection element next to it, and finding something in the typeahead adds via javascript something to the collection, using the collection's prototype. In fact I'm thinking we need to abstract that code away, the exact same code should work in all those cases, up to some customization via data parameters.

**Comment by altermattw** [comment]
This might be the relevant section of TagController.  I just tried a git checkout from before I started messing with the Tag Edit screen and it persists just fine.  So, I'm looking to see what might be different between the two forms.  One big difference is that the original form inputs were `<select>` elements, and I replaced those with `<input type="hidden">`.  Would that break it?
```
$tm = $this->get('model.tag'); // Not sure where it collects all the form data. Here?
$tag = $isNew ? $tm->make() : $tm->find($tag_id);
$originalOrdered = new ArrayCollection();
foreach ($tag->getOrderedQuestions() as $oq) {
    $originalOrdered->add($oq); // This loads up $originalOrdered with the ordered 
     // questions currently in the database.
}
...
$updatedQuestions = $tag->getOrderedQuestions(); // How is this different from
    // $originalOrdered?  Presumably, $updatedQuestions should be the new questions 
    // that are on the user's screen.
if ($form->isSubmitted()) {
    if ($form->get('cancel')->isClicked()) {
    ...
    }
if ($form->isValid()) {
    foreach ($originalOrdered as $oq) { // cycle through each of the previously existing questions
        if (!$updatedQuestions->contains($oq)) { // if the new questions don't contain one of the
     // original questions
            $tm->em->remove($oq); // then remove the existing question
        }
    }
    $tm->em->flush();

    $this->get('model.orderedQuestion')->fixOrder($updatedQuestions);
    $tm->em->flush();

    return $this->redirectToRoute('admin_tag_list');
}
```

**Comment by skiadas** [comment]
So, the data is collected in the line `$tag = $isNew ? $tm->make() : $tm->find($tag_id);` it either creates a new blank tag entity or it looks one up using `find`.

Then the tag entity contains a list of orderedQuestions. We need to store that list to compare it against the resulting list.

Now the  `$updatedQuestions = $tag->getOrderedQuestions();` seems wrong, it should have been the data from the `$form->getData()` instead, but that won't happen until a bit later. So I am not sure, maybe it's some leftover code.

So the key idea here, whether it's implemented properly or not, is that an "orderedQuestion" item represents a specific location in the list of questions in a specific tag. So if a tag has 5 questions in it, there will be exactly 5 orderedQuestion items for that tag. The combination of a tag and order number is a primary key for the orderedQuestions.

The "content" of an orderedQuestion is the actual question that's there, and that may change when we shuffle things around. Also when we remove items, we need to end up with fewer orderedQuestions altogether, and when we add items we need to create new orderedQuestions. And in any case we might need to do some careful renumbering, for instance if we remove the question numbered 2, the others will not change number in the form, but we do need to change them on the backend, so that the third becomes 2nd, the 4th becomes 3rd and so on. That's what "fixOrder" does.

**Comment by skiadas** [comment]
Ok I found out one part of the "confusion". Remember how I was wondering why you had to dig a bit deep into the orderedQuestion item to find what you were looking for? It is because what you were accessing was not an orderedQuestion entity, but an orderedQuestionType object. And *that* we need to customize, because by default it's not doing the right things really. But that is the source of the problems.

**Comment by skiadas** [comment]
@altermattw I made a first step at isolating the fragment code, take a look at the tagFormHaris branch. I didn't work on the Javascript functionality at all yet, nor changed the AddQuestion button, but the list should be rendering quite nicely. And you'll see the prototype is automatically generated. (Basically the prototype also uses the same fragment used for the rows, but it passed it a "value" of null, so we use "defaults" to give it the placeholder strings it needs to have).

**Comment by skiadas** [comment]
Don't try to "save" a tag right now, it will lose the ordered questions you currently have. Needs some more work.

**Comment by altermattw** [comment]
The tagFormHaris branch looks great!  The typeahead styling looks better, too.  One problem with the typeahead on that page, though, is that the last row of the suggested questions (under "Add Question") is hidden under the red page footer.

**Comment by altermattw** [comment]
I removed the footer so that the dropdown options weren't hidden behind it.  We don't really need the footer now that I've got a contact page.
The Add Item function isn't working now, though.  I think it's because of the problem you mentioned with Typeahead serving two masters.  An error occurs in typeaheadCollection.js here:
```
$insertedEl = $(prototype.replace(/__name__/g, count)
            .replace(/__value__/g, obj.value));
```
I think the problem is that there is no data-prototype attribute in this case.  Should we create two separate TypeAhead methods?
